cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/UserTaskManager.java

```diff
@@ -143,69 +144,87 @@ private static String httpServletRequestToString(HttpServletRequest request) {
+      Function<String, OperationFuture> function, int step, boolean asyncComplete) {
     UUID userTaskId = getUserTaskId(httpServletRequest);
-    List<OperationFuture> operationFutures = getFuturesByUserTaskId(userTaskId, httpServletRequest);
+    UserTaskInfo userTaskInfo = getUserTaskByUserTaskId(userTaskId, httpServletRequest);
 
-    if (operationFutures != null) {
+    if (!asyncComplete && userTaskInfo != null) {
+      throw new IllegalArgumentException(
+          String.format("Sync operation should not have %s", USER_TASK_HEADER_NAME));
+    }
+
+    if (userTaskInfo != null) {
       LOG.info("Fetch an existing UserTask {}", userTaskId);
       httpServletResponse.setHeader(USER_TASK_HEADER_NAME, userTaskId.toString());
-      if (step < operationFutures.size()) {
-        return operationFutures;
-      } else if (step == operationFutures.size()) {
+      if (step < userTaskInfo.futures().size()) {
+        return userTaskInfo.futures();
+      } else if (step == userTaskInfo.futures().size()) {
         LOG.info("Add a new future to existing UserTask {}", userTaskId);
-        return insertFuturesByUserTaskId(userTaskId, function, httpServletRequest);
+        return insertFuturesByUserTaskId(userTaskId, function, httpServletRequest).futures();
       } else {
         throw new IllegalArgumentException(
-            String.format("There are %d steps in the session. Cannot add step %d.", operationFutures.size(), step));
+            String.format("There are %d steps in the session. Cannot add step %d.", userTaskInfo.futures().size(), step));
       }
     } else {
-      if (httpServletRequest.getHeader(USER_TASK_HEADER_NAME) != null) {
-        // request provides user_task_header that user tasks doesn't exist then just throw exception
-        String userTaskIdFromRequest = httpServletRequest.getHeader(USER_TASK_HEADER_NAME);
-        throw new IllegalArgumentException(
-            String.format("UserTask %s is an invalid %s", userTaskIdFromRequest, USER_TASK_HEADER_NAME));
-      }
-
+      ensureUserTaskHeaderNotPresent(USER_TASK_HEADER_NAME, httpServletRequest);
       if (step != 0) {
         throw new IllegalArgumentException(
             String.format("There are no step in the session. Cannot add step %d.", step));
       }
-
-      SessionKey sessionKey = new SessionKey(httpServletRequest);
       userTaskId = _uuidGenerator.randomUUID();
-      LOG.info("Create a new UserTask {} with SessionKey {}", userTaskId, sessionKey);
-      operationFutures = insertFuturesByUserTaskId(userTaskId, function, httpServletRequest);
-
-      synchronized (_sessionKeyToUserTaskIdMap) {
-        _sessionKeyToUserTaskIdMap.put(sessionKey, userTaskId);
+      userTaskInfo = insertFuturesByUserTaskId(userTaskId, function, httpServletRequest);
+      // Only create user task id to session mapping for async request
+      if (asyncComplete) {
+        createSessionKeyMapping(userTaskId, httpServletRequest);
```

Comment:
Same as above, is there a particular reason that sync request should not have a session?
User can still use session+URL to retrieve the previous result

Reply:
Similar to above, if sync requests are meant to return right away, we don't need to keep them in session and hope to retrieve their result using UUID.